docs/architecture/adr-020-protobuf-transaction-encoding.md

```diff
@@ -135,40 +166,155 @@ Then, each module's client handler will at the minimum accept a `Marshaler` inst
+`sdk.Msg` that uses `Any`.** Convenience gRPC/REST methods are described later on
+that allow the generic module-level `Msg` types to be used for transaction
+submission, converting them to the app-level encoding `Msg`s that use `oneof`
+behind the scenes.
+
+In order to smoothly allow for conversion between _signing_ and _encoding_
+`Msg`s, modules that use interface types should implement the following
+interface:
+
+```go
+type InterfaceMsgEncoder {
+  GetSigningMsg(sdk.Msg) (sdk.Msg, error)
+  GetEncodingMsg(sdk.Msg) (sdk.Msg, error)
 }
 ```
 
-Note that the implementation of `MsgSubmitProposalI` can be simplified by
-using an embedded base struct which implements most of that interface - in this
-case `MsgSubmitProposalBase`.
+`GetSigningMsg` should type switch over the `sdk.Msg`s that use interfaces and
+convert `Msg`s in the encoding format (that use `oneof`)
+to the signing version (that uses `Any`). `GetEncodingMsg` should
+convert `Msg`s used for signing (using `Any`) to those used for encoding
+(using `oneof`). Ex:
+
+```go
+// x/gov/module.go
+var _ InterfaceMsgEncoder = AppModule{}
+
+type AppModule struct {
+  ...
+  newEncodingMsg fn(MsgSubmitProposalBase, Content) (MsgSubmitProposalI, error)
+}
+
+func (am AppModule) GetSigningMsg(msg sdk.Msg) (sdk.Msg, error) {
+	switch msg := msg.(type) {
+	case MsgSubmitProposalI:
+        return NewMsgSubmitAnyProposal(msg.GetBase(), msg.GetContent()), nil
+	default:
+		return msg, nil
+	}
+}
+
+func (am AppModule) GetEncodingMsg(sdk.Msg) (sdk.Msg, error) {
+	switch msg := msg.(type) {
+	case MsgSubmitProposalI:
+        return am.newEncodingMsg(msg.GetBase(), msg.GetContent())
+	default:
+		return msg, nil
+	}
+}
+```
+
+Because apps will know how to convert signing `Msg`s to encoding `Msg`s,
+signing `Msg`s can and should be used for transaction submission against a
+helper function (for CLI methods) or RPC method that does the actually encoding.
+
+### Generic gRPC/REST Transaction Service
+
+The usage of `Any` by clients as described above allows us to provide a generic
+app-independent transaction service (leveraging the generic `AnyTransaction` type
+described above) which provides similar functionalities as the existing REST
+server. This service will have a gRPC schema roughly as follows with REST
+endpoints provided by grpc-gateway:
+
+```go
+service TxService {
+    // Get a Tx by hash
+    rpc QueryTx (QueryTxRequest) returns (QueryTxResponse) {
+      option (google.api.http) = {
+        get: "/txs/{hash}"
+      };
+    }
+    
+    // Search transactions
+    rpc QueryTxs (QueryTxsRequest) returns (QueryTxResponse) {
+      option (google.api.http) = {
+        get: "/txs"
+        body: "*"
+      };
+    }
+ 
+    // Broadcast a signed tx
+    rpc BroadcastTx (BroadcastTxRequest) returns (BroadcastTxResponse) {
+      option (google.api.http) = {
+        post: "/txs"
+        body: "*"
+      };
+    }
+}
+
+message BroadcastTxRequest {
+    AnyTransaction tx = 1;
```

Comment:
Where would this get the .proto files for all the possible message types get from, which are necessary to decode `AnyTransaction` from JSON and into protobuf? Download from the URLs?

Reply:
Developers would get them from git repos and run code generation to javascript, etc. Likely someone will create npm packages with the generated code